Oidc Auth for Azure steps

IsaacCalligeros95 opened the pull request:
This PR supports AzureOidc Accounts to Calamari for all Azure steps **excluding CloudService & ServiceFabric**. The associated Server changes can be found [here](https://github.com/OctopusDeploy/OctopusDeploy/pull/20042). 

As a part of this change, ServicePrincipal was implemented mostly per CalamariFlavour basis. This has been consolidated into the _Calamari.CloudAccounts_ repository, which is now shared across most Azure Calamari repositories. This adds some unnecessary overhead to the Calamari Bundles for some of these flavours, but the included Nuget Packages are relatively small ~1MB for the largest, and given the additional duplication introduced for AzureOidc Accounts looks to be worth the cost of consolidating.

[See Server PR for test cases](https://github.com/OctopusDeploy/OctopusDeploy/pull/20042).)
[sc-57210}

IsaacCalligeros95 commented on source/Calamari.Azure/ServicePrincipalAccount.cs:
```diff
@@ -7,7 +7,7 @@
 
 namespace Calamari.Azure
 {
-    public class ServicePrincipalAccount
```
There's no actual shared base project for Calamari Azure steps. This one is specifically for Kubernetes cloud target discovery and is only included in the base Calamari Project. I'd like to pull the ServicePrincipalAccount, AzureOidcAccount and ideally all accounts into Calamari.Common which is used in the majority of steps. I have not done this as a part of this PR, at least not yet any feedback or hesitations on making this change?

IsaacCalligeros95 replied:
An initial cut at this required bringing in the following packages to Calamari.Common. Alterntivly these could just be extension methods in the individual flavours that requite these (AzureWebApp & Calamari.Azure) It's likely worthwhile doing this but I'm holding off for feedback

`using Microsoft.Azure.Management.Fluent;
using Microsoft.Azure.Management.ResourceManager.Fluent.Authentication;
using Microsoft.Rest;`

IsaacCalligeros95 commented on source/Calamari.AzureAppService/Calamari.AzureAppService.csproj:
```diff
@@ -17,15 +17,15 @@
     <PackageReference Include="Microsoft.Azure.Management.Fluent" Version="1.37.1" />
-    <PackageReference Include="Microsoft.Azure.Management.Websites" Version="3.1.1" />
-    <PackageReference Include="Microsoft.Identity.Client" Version="4.48.1" />
+    <PackageReference Include="Microsoft.Azure.Management.Websites" Version="4.0.0" />
```
3.1.1 returns a 404 on Access Tokens Generated through OAuth flows

IsaacCalligeros95 commented on source/Calamari.AzureAppService/Behaviors/TargetDiscoveryBehaviour.cs:
```diff
@@ -109,7 +111,7 @@ public async Task Execute(RunningDeployment runningDeployment)
         private void WriteTargetCreationServiceMessage(
             AzureResource resource,
-            TargetDiscoveryContext<AccountAuthenticationDetails<ServicePrincipalAccount>> context,
+            TargetDiscoveryContext<AccountAuthenticationDetails<AzureServicePrincipalAccount>> context,
```
Oidc Accounts are not supported in Calamari yet

IsaacCalligeros95 commented on source/Calamari.CloudAccounts/Calamari.CloudAccounts.csproj:
```diff
@@ -22,7 +22,10 @@
       <PackageReference Include="AWSSDK.Core" Version="3.3.104.14" />
       <PackageReference Include="AWSSDK.SecurityToken" Version="3.3.4.34" />
-      <PackageReference Include="Newtonsoft.Json" Version="9.0.1" />
+      <PackageReference Include="Newtonsoft.Json" Version="10.0.3" />
```
Net452 limits to version 10.0.3, rather than the later 13.0.3 used in other projects

IsaacCalligeros95 commented on source/Calamari.AzureAppService/Behaviors/TargetDiscoveryBehaviour.cs:
```diff
@@ -1,22 +1,21 @@
 using Calamari.Common.Plumbing.Pipeline;
 using Calamari.Common.Plumbing.ServiceMessages;
 using Calamari.Common.Plumbing.Variables;
 using Newtonsoft.Json;
```
Switched to Newtonsoft.Json deserialization as Calamari.CloudAccounts builds net452 and System.Text.Json is only available from net461 on. Newtonsoft is case insensitive by default which matches the existing behaviour

benPearce1 commented on source/Calamari.AzureAppService/AppSettingsManagement.cs:
```diff
@@ -98,7 +98,7 @@ public static class AppSettingsManagement
             var results = await client.GetStringAsync(targetUrl);
 
-            var output = JsonConvert.DeserializeObject<appSettingNamesRoot>(results).properties.appSettingNames ??
+            var output = JsonConvert.DeserializeObject<appSettingNamesRoot>(results)?.properties.appSettingNames ??
```
did this cause a problem when you were testing?

IsaacCalligeros95 replied:
NewtonSoft.Json returns a nullable string compared to System.Text.Json returning a string. I had to swap to NewtonSoft because of the Calamari.CloudAccounts net452 build

IsaacCalligeros95 commented on source/Calamari.AzureAppService/Azure/AzureClient.cs:
```diff
@@ -1,87 +1,85 @@
 {
     internal static class AzureClient
     {
-        public static IAzure CreateAzureClient(this ServicePrincipalAccount servicePrincipal)
-        {
-            var environment = new AzureKnownEnvironment(servicePrincipal.AzureEnvironment).AsAzureSDKEnvironment();
-            var credentials = SdkContext.AzureCredentialsFactory.FromServicePrincipal(servicePrincipal.ClientId,
-                servicePrincipal.Password, servicePrincipal.TenantId, environment
-            );
-
-            // Note: This is a tactical fix to ensure this Sashimi uses the appropriate web proxy
-#pragma warning disable DE0003
-            var client = new HttpClient(new HttpClientHandler {Proxy = WebRequest.DefaultWebProxy});
-#pragma warning restore DE0003
-
-            return Microsoft.Azure.Management.Fluent.Azure.Configure()
-                            .WithHttpClient(client)
-                            .Authenticate(credentials)
-                            .WithSubscription(servicePrincipal.SubscriptionNumber);
-        }
```
Pushed down to the Calamari.CloudAccounts, each of AzureServicePrincipal and AzureOidcAccount have their implementation of this method with a shared interface.

benPearce1 commented on source/Calamari.AzureAppService/Calamari.AzureAppService.csproj:
```diff
@@ -17,19 +17,20 @@
       <PrivateAssets>all</PrivateAssets>
       <IncludeAssets>runtime; build; native; contentfiles; analyzers</IncludeAssets>
     </PackageReference>
+    <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
```
Was there a specific reason for swapping to NewtonSoft?
Sometimes, serialisation libraries can give you different results if you don't setup the configuration right

IsaacCalligeros95 replied:
Calamari.CloudAccounts builds net452 and System.Text.Json is only available from net461 on. Newtonsoft is case insensitive by default, which matches the existing behaviour we specified in System.Text.Json in the `source/Calamari.AzureAppService/Behaviors/TargetDiscoveryBehaviour.cs`

benPearce1 replied:
sorry, just saw your response from earlier

benPearce1 commented on source/Calamari.AzureAppService/Behaviors/AzureAppServiceBehaviour.cs:
```diff
@@ -0,0 +1 @@
+﻿
```
empty file. should be removed, doesn't have to be this pr

benPearce1 commented on source/Calamari.AzureScripting/AzureContextScriptWrapper.cs:
```diff
@@ -62,17 +63,28 @@ public AzureContextScriptWrapper(IVariables variables, ICalamariFileSystem fileS
                 {
-                    SetOutputVariable("OctopusUseServicePrincipal", bool.TrueString);
+                    SetOutputVariable("OctopusAzSpOrOidc", bool.TrueString);
                     SetOutputVariable("OctopusAzureADTenantId", variables.Get(SpecialVariables.Action.Azure.TenantId)!);
                     SetOutputVariable("OctopusAzureADClientId", variables.Get(SpecialVariables.Action.Azure.ClientId)!);
-                    variables.Set("OctopusAzureADPassword", variables.Get(SpecialVariables.Action.Azure.Password));
+
+                    if (variables.Get(SpecialVariables.Account.AccountType) == "AzureServicePrincipal")
+                    {
+                        variables.Set("OctopusUseServicePrincipal", bool.TrueString);
+                        variables.Set("OctopusAzureADPassword", variables.Get(SpecialVariables.Action.Azure.Password));
+                    }
+                    else
+                    {
+                        variables.Set("OctopusOpenIdJwt", variables.Get(AccountVariables.Jwt));
+                    }
```
The variables being set here are a bit confusing. For an OIDC account we get `OctopusAzSpOrOidc` set to `true`, but for a password based service principal account we get `OctopusAzSpOrOidc` and `OctopusUseServicePrincipal` set to `true`.
Don't think it is worth changing, but food for thought. Also, not a massive fan of the shortened `OctopusAzSpOrOidc`, would prefer to spell it all out, `OctopusAzureServicePrincipalOrOidc` 🤷

IsaacCalligeros95 replied:
I'd meant to tidy this up, I think the scripts mostly had this removed but missed the wrapper. Thanks for the call out

benPearce1 reviewed (approved):
This is good. There are some really minor points for improvement, but I would like to merge this and review later when we/if have time.

IsaacCalligeros95 commented:
> This is good. There are some really minor points for improvement, but I would like to merge this and review later when we/if have time.

Addressed comments in branch for Terraform [work here](https://github.com/OctopusDeploy/Calamari/commit/c20faf61088d50a9e61e786f798c0b565862ab79)
